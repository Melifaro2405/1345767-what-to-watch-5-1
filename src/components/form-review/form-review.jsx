import React from "react";
import PropTypes from "prop-types";
import {sendReview} from "../../serviсes/api-actions";
import {connect} from "react-redux";

const RATE_STARS = [1, 2, 3, 4, 5];
const RATING_MULTIPLIER = 2;

const FormReview = ({
  isActive,
  isLoading,
  isError,
  changeIsLoading,
  changeIsError,
  filmID,
  onSubmit,
  text,
  rating,
  onChangeText,
  onChangeRating
}) => {

  const handleSubmitReview = (evt) => {
    evt.preventDefault();

    changeIsLoading(true);
    onSubmit({
      id: filmID,
      rating: Number(rating) * RATING_MULTIPLIER,
      comment: text
    })
    .catch(() => {
      changeIsLoading(false);
      changeIsError(true);
    });
  };

  return (
    <form onSubmit={handleSubmitReview} action="#" className="add-review__form">
      <div className="rating">
        <div className="rating__stars" onChange={onChangeRating}>
          {RATE_STARS.map((star) =>
            <React.Fragment key={star}>
              <input
                className="rating__input"
                id={`star-${star}`}
                type="radio"
                name="rating"
                value={star}
                defaultChecked={Number(rating) === star}
              />
              <label
                className="rating__label"
                htmlFor={`star-${star}`}
              >
                {`Rating ${star}`}
              </label>
            </React.Fragment>
          )}
        </div>
      </div>

      <div className="add-review__text">
        <textarea
          minLength="50"
          maxLength="400"
          value={text}
          onChange={onChangeText}
          className="add-review__textarea"
          name="review-text"
          id="review-text"
          placeholder="Review text"
        ></textarea>
        <div className="add-review__submit">
          <button
            className="add-review__btn"
            type="submit"
            disabled={!isActive || isLoading}
          >Post
          </button>
        </div>
      </div>
      {isError && <p style={{fontWeight: `bold`, textAlign: `center`, color: `red`}}>
        ОШИБКА ОТПРАВКИ ФОРМЫ
      </p>}
    </form>
  );
};

FormReview.propTypes = {
  filmID: PropTypes.number.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChangeText: PropTypes.func.isRequired,
  onChangeRating: PropTypes.func.isRequired,
  text: PropTypes.string.isRequired,
  rating: PropTypes.string.isRequired,
  isActive: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isError: PropTypes.bool.isRequired,
  changeIsLoading: PropTypes.func.isRequired,
  changeIsError: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch) => ({
  onSubmit(comment) {
    return dispatch(sendReview(comment));
  }
});

export {FormReview};
export default connect(null, mapDispatchToProps)(FormReview);
