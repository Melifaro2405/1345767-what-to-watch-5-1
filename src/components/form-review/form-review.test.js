import React from 'react';
import renderer from 'react-test-renderer';
import {FormReview} from './form-review';

const noop = () => {};

describe(`Should FormReview render correctly`, () => {
  test.each([
    [`all false`, false, false, false],
    [`isActive true, isLoading & isError false`, true, false, false],
    [`isError true, isActive & isLoading false`, false, false, true],
    [`isLoading true, isActive & isError false`, false, true, false],
    [`isActive & isLoading true, isError false`, true, true, false],
    [`isActive & isError true, isLoading false`, true, false, true],
    [`isLoading & isError true, isActive false`, false, true, true],
    [`all true`, true, true, true]
  ])(`with %s`, (_expected, isActive, isLoading, isError) => {
    const tree = renderer
      .create(<FormReview
        isActive={isActive}
        isLoading={isLoading}
        isError={isError}
        filmID={1}
        text={`test`}
        rating={`3`}
        changeIsError={noop}
        changeIsLoading={noop}
        onChangeText={noop}
        onChangeRating={noop}
        onSubmit={noop}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
