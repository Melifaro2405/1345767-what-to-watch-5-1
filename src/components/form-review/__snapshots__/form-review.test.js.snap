// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FormReview render correctly with all false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;

exports[`Should FormReview render correctly with all true 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
  <p
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    ОШИБКА ОТПРАВКИ ФОРМЫ
  </p>
</form>
`;

exports[`Should FormReview render correctly with isActive & isError true, isLoading false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={false}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
  <p
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    ОШИБКА ОТПРАВКИ ФОРМЫ
  </p>
</form>
`;

exports[`Should FormReview render correctly with isActive & isLoading true, isError false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;

exports[`Should FormReview render correctly with isActive true, isLoading & isError false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={false}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;

exports[`Should FormReview render correctly with isError true, isActive & isLoading false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
  <p
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    ОШИБКА ОТПРАВКИ ФОРМЫ
  </p>
</form>
`;

exports[`Should FormReview render correctly with isLoading & isError true, isActive false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
  <p
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    ОШИБКА ОТПРАВКИ ФОРМЫ
  </p>
</form>
`;

exports[`Should FormReview render correctly with isLoading true, isActive & isError false 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
      onChange={[Function]}
    >
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-1"
        name="rating"
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-2"
        name="rating"
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        className="rating__input"
        defaultChecked={true}
        id="star-3"
        name="rating"
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-4"
        name="rating"
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        className="rating__input"
        defaultChecked={false}
        id="star-5"
        name="rating"
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="test"
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;
